cpp/ql/lib/semmle/code/cpp/rangeanalysis/new/RangeAnalysis.qll

```diff
@@ -0,0 +1,95 @@
+private import semmle.code.cpp.rangeanalysis.new.internal.semantic.analysis.RangeAnalysis
+private import cpp
+private import semmle.code.cpp.ir.IR
+private import semmle.code.cpp.controlflow.IRGuards
+private import semmle.code.cpp.rangeanalysis.new.internal.semantic.SemanticExpr
+private import semmle.code.cpp.rangeanalysis.new.internal.semantic.SemanticExprSpecific
+private import semmle.code.cpp.rangeanalysis.new.internal.semantic.analysis.Bound as IRBound
+private import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/**
+ * Holds if e is bounded by `b + delta`. The bound is an upper bound if
+ * `upper` is true, and can be traced baack to a guard represented by `reason`.
+ */
+predicate bounded(Expr e, Bound b, float delta, boolean upper, Reason reason) {
+  exists(SemanticExprConfig::Expr semExpr |
+    semExpr.getUnconverted().getUnconvertedResultExpression() = e
+    or
+    semExpr.getConverted().getConvertedResultExpression() = e
+  |
+    semBounded(semExpr, b, delta, upper, reason)
+  )
+}
```

Comment:
Instead of lumping the converted and the unconverted expressions together into a single predicate, how about we have two predicates:
```ql
/**
 * Holds if e is a non-conversion expression bounded by `b + delta`. The bound is an
 * upper bound if `upper` is true, and can be traced baack to a guard represented
 * by `reason`.
 */
predicate bounded(Expr e, Bound b, float delta, boolean upper, Reason reason) { ... }

/**
 * Holds if e is an expression bounded by `b + delta`. The `Expr` may be a conversion.
 * The bound is an upper bound if `upper` is true, and can be traced baack to a guard
 * represented by `reason`.
 */
predicate convertedBounded(Expr e, Bound b, float delta, boolean upper, Reason reason) { ... }
```
The first predicate would then use `semExpr.getUnconverted().getUnconvertedResultExpression() = e`, and the second would use `semExpr.getConverted().getConvertedResultExpression() = e`.

How does that sound?

Reply:
Sounds good to me.